components/org.wso2.carbon.security.caas/src/main/java/org/wso2/carbon/security/caas/api/module/UsernamePasswordLoginModule.java

```diff
@@ -102,18 +111,51 @@ public boolean login() throws LoginException {
                     CarbonSecurityErrorMessages.UNSUPPORTED_CALLBACK_EXCEPTION.getDescription(), e);
         } catch (IOException e) {
             throw new CarbonSecurityServerException(CarbonSecurityErrorMessages.CALLBACK_HANDLE_EXCEPTION.getCode(),
-                                                    CarbonSecurityErrorMessages.CALLBACK_HANDLE_EXCEPTION
-                                                            .getDescription(), e);
+                    CarbonSecurityErrorMessages.CALLBACK_HANDLE_EXCEPTION.getDescription(), e);
         }
 
         username = usernameCallback.getName();
         password = passwordCallback.getPassword();
 
-
+        uncommittedUser = validateUserPassword(username, password);
+        success = uncommittedUser != null;
 
         //TODO Add Audit logs CARBON-15870
-        success = true;
-        return true;
+        return success;
+    }
+
+    private User validateUserPassword(String userName, char[] credentials)
+            throws CarbonSecurityServerException, CarbonSecurityAuthenticationException {
+        if (credentials == null) {
+            return null;
+        }
+        User user = CarbonSecurityUtils.getUser(userName);
+        if (user == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("User not found for userName: %s. Failing the authentication." + userName);
+            }
+            throw new CarbonSecurityAuthenticationException();
+        } else {
+            byte[] plainUserPassword = decoder.decode(user.getPassword());
+            if (!Arrays.equals(plainUserPassword, toBytes(credentials))) {
+                if (log.isDebugEnabled()) {
+                    log.debug(
+                            "Password did not match with the configured user, userName: %s, Failing the authentication."
+                                    + userName);
+                }
+                throw new CarbonSecurityAuthenticationException();
+            }
+        }
+        return user;
+    }
+
+    private byte[] toBytes(char[] chars) {
+        CharBuffer charBuffer = CharBuffer.wrap(chars);
+        ByteBuffer byteBuffer = Charset.forName("UTF-8").encode(charBuffer);
```

Comment:
We can use StandardCharsets.UTF_8 (which comes with java7 onwards) instead of Charset.forName("UTF-8") also